contract/src/lib.rs

```diff
@@ -388,6 +402,50 @@ impl Contract {
             false
         }
     }
+
+    // Current GC implementation is doing full travers of the blocks starting from the main chain
+    // tip. We can optimize this, by storing the height of a current genesis block in our state.
+    pub fn run_mainchain_gc(&mut self, batch_size: u8) {
+        let initial_blockheader = self
+            .headers_pool
+            .get(&self.mainchain_initial_blockhash)
+            .expect("initial blockheader must be in a header pool");
+
+        let tip_blockheader = self
+            .headers_pool
+            .get(&self.mainchain_tip_blockhash)
+            .expect("tip blockheader must be in a header pool");
+
+        let amount_of_headers_we_store =
+            tip_blockheader.block_height - initial_blockheader.block_height;
+
+        if amount_of_headers_we_store > self.gc_threshold as u64 {
+            let total_amount_to_remove = amount_of_headers_we_store - self.gc_threshold as u64;
+            let selected_amount_to_remove =
+                std::cmp::min(total_amount_to_remove, batch_size as u64);
+
+            let start_removal_height = initial_blockheader.block_height;
+            let end_removal_height = initial_blockheader.block_height + selected_amount_to_remove;
+
+            for height in start_removal_height..end_removal_height {
+                let blockhash = self
+                    .mainchain_height_to_header
+                    .get(&height)
+                    .expect("target height should exist")
+                    .to_string();
```

Comment:
The type is already a string; you don't need `to_string` here.

Reply:
rewrote in a different way